オブジェクト指向プログラミング課題（1st draft）

makiichikawa opened the pull request:
オブジェクト指向のどういう性質を使って書き換えたのかの説明についてはコミットメッセージに記載しました🙏

makiichikawa committed:
isbnの取得をコントローラからモデルへ移動
- モデルのパブリックインターフェースとしてはコントローラから細い指示を減らすことが望ましいため
- なぜなら、今後isbnになんらかの変更が生じた場合、コントローラとモデル両方に修正する必要があり、安定したパブリックインターフェースとしては言いがたくなってしまうため

makiichikawa committed:
インスタンス変数を直接参照しないように変更（データの隠蔽）
- インスタンス変数の管理値変更に影響を受けてしまうため
- 著者オブジェクトを丸ごと渡すことができるため

makiichikawa committed:
createメソッド内にあるパラメータ抽出部分をプライベートメソッドとして切り出す
- createメソッドはbookインスタンスを登録し、ビューをレンダリングする役割がある
- パラメータの抽出は余計な責務と考え、プライベートメソッドに切り出し
- 将来、パラメータが追加される可能性があるので、前もって隔離しておく

yoshitsugu commented on opp_practice.rb:
```diff
@@ -17,11 +21,20 @@ def index
+    author = Author.create!(author_params)
+    book = Book.new(user_params)
+    book.author = author
+    book.save!
```
Controllerから見たとき、 `本を作成したい` のであって、BookとAuthorが関連テーブルである、というのは特に主要な関心事ではないですよね。Bookの中にロジックを隠蔽できないか、考えてみてください！

yoshitsugu reviewed (commented):
コメントしました！

makiichikawa commented:
@yoshitsugu 

確認ありがとうございます！
また、考えて、修正でき次第連絡します🙇

makiichikawa committed:
コードに誤りがあったため、修正

makiichikawa committed:
Controllerから不要な責任を取り除く
- Controllerの関心事は本を作成するであり、BookとAuthorの紐付けではない
- Bookメソッドに責任を委ねる

makiichikawa commented:
@yoshitsugu 

指摘取り込み完了しましたので、確認のほどよろしくお願い致します🙇

yoshitsugu commented on opp_practice.rb:
```diff
@@ -1,6 +1,10 @@
 # 本モデル
 class Book < ApplicationRecord
   belongs_to :author
+  before_save do
+    # 本にはISBNという番号が割りふられる。ここではその番号を自動で取得してくれるサービスがあるとする
+    self.isbn = Faraday.get('/isbn_publish_service')
+  end
```
まあこれでもいいんですが、基本的にRailsのコールバックはどこで設定されたのかわかりづらくなるので、あまり使いたくはないですね。
`isbn` と `author` も含めて保存するようなメソッドを作ってそちらで明示的に作るようにしましょう。

yoshitsugu commented on opp_practice.rb:
```diff
@@ -17,11 +21,18 @@ def index
-    isbn = Faraday.get('/isbn_publish_service')
-    book = Book.create!(author_id: author.id, title: params[:book][:title], published_at: params[:book][:published_at], isbn: isbn)
+    book = Book.create(book_params)
+    book.create_author(author_params)
```
結局これも `author` 用のメソッドを別に呼び出さないといけない、という知識が必要になりますよね。
それより上に書いた通り、よしなにやってくれるメソッドを1つ作ったほうがよいかと思います。

yoshitsugu reviewed (commented):
@makiichikawa 
コメントしました！

makiichikawa commented:
@yoshitsugu 

コメントありがとうございます🙏
また、対応完了したら連絡いたします🙇

makiichikawa committed:
- BookオブジェクトをAuthorオブジェクトとともに作成するクラスメソッドを用意
  - BookControllerとしてはBookオブジェクトを作成して、Viewに返せればいいため
  - 関連先のAuthorオブジェクトまで作成することは責任の範囲外のため

makiichikawa committed:
保存後の処理をcreate_book_with_authorのメソッドに移動
  - コールバック処理は設定場所が分かりずらいため

makiichikawa commented:
@yoshitsugu

以下いただいた指摘を踏まえて変更しました。

https://github.com/makiichikawa/oop_practice/pull/1#discussion_r1045410850
https://github.com/makiichikawa/oop_practice/pull/1#discussion_r1045411323

お手隙の際に確認お願いいたします🙇

yoshitsugu commented on opp_practice.rb:
```diff
@@ -1,10 +1,18 @@
+    def create_book_with_author(args)
+      # 本にはISBNという番号が割りふられる。ここではその番号を自動で取得してくれるサービスがあるとする
+      isbn = Faraday.get('/isbn_publish_service')
+      Author.create(name: args[:author_name]).books.create(name: args[:name], title: args[:title], isbn: isbn)
```
- これはオブジェクト指向というよりRailsの話なんですが、 `create` だとバリデーションなどの失敗時に `false` になるだけなので、 `create!` を使うか、 `if` 文などで失敗したときのハンドリングが必要になります。
- これだとAuthorが同じbookが作られるときに毎回Authorを作ってしまうので、 `create_or_find_by!` などが必要になるかと思います。
- https://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-create_or_find_by-21

yoshitsugu replied:
あと、2種類のレコードを触っているので `transaction` が必要ないか、考えてみてください。

makiichikawa replied:
@yoshitsugu 

ありがとうございます🙏
Railsでの挙動を確認した上で修正します。
終わり次第連絡します🙏

makiichikawa replied:
@yoshitsugu

上記指摘取り込み完了しています。

transactionについては必要ないと考えました。
authorを保存し、bookが保存されないケースに備えてtransaction必要になると思いますが、
今回の修正で、book、author両方ともにバリデーションをかけているので、上記のケースは起こらないと考えました。

お手隙の際にご確認のほどよろしくお願いいたします🙇

yoshitsugu replied:
@makiichikawa たとえば、もしバリデーション設定漏れがあって、DBの制約のほうが厳しかったりすると、 エラーになるケースはあると思います。なので、理由としては不十分かなと思います。
一方で、ゴミAuthorデータができることを許容するなら、 `create_or_find_by!` 使っている以上、二回目以降はAuthorをfindするだけなので、transactionなくてもあまり危険性はないです。
バリデーションがあるからtransactionいらない、という判断はできない、という点だけ注意してください。

makiichikawa replied:
> しバリデーション設定漏れがあって、DBの制約のほうが厳しかったりすると、 エラーになるケースはある
> バリデーションがあるからtransactionいらない

承知しました、ご指摘ありがとうございます🙏

yoshitsugu reviewed (commented):
ほぼOKで、Rails的な話になりますが、コメントしています。

makiichikawa committed:
- Authorに一意制約をつける
- Authorの保存前にバリデーションを実施
- createからcreate_or_find_by!に変更

makiichikawa committed:
Bookモデルの属性に一部誤りがあったため修正
- nameではなく、ただしくはpublish_at

makiichikawa committed:
- Bookモデルにバリデーションを追加
- createからcreate!に変更

yoshitsugu reviewed (approved):
OKですが、一点だけコメントしています。
